### ISSUE TITLE
dhall-docs test golden files can't be opened with a browser

### ISSUE THREAD

**Issue by german1608** [dhall-lang/dhall-haskell#1967]
The reason for this is because HTML requires `script` tags to be always closed even when there is no content i.e. the following:

```html
<script src="..." ... />
```
is invalid, it should be written as:

```html
<script src="..." ... ></script>
```

It looks like [`HaXML`](https://hackage.haskell.org/package/HaXml) is not taking that into account when rendering as HTML

**Comment by german1608** [comment]
... found at #1966

**Comment by sjakobi** [comment]
Fixing `HaXML` seems tricky – it doesn't even seem to have a proper bug tracker. (There's a bug report [here](https://github.com/haskell-infra/hackage-trustees/issues/273) though.)

So instead of relying on `HaXML`, could we possibly do something else to make the golden HTML files somewhat readable and, more importantly, make their diffs readable?

How about simply replacing each space character with a newline, for example?

**Comment by sjakobi** [comment]
It might also be helpful to file a pretty-printing feature request for `lucid` – maybe that would just get us a tool recommendation for a different HTML prettifier.

**Comment by german1608** [comment]
> It might also be helpful to file a pretty-printing feature request for `lucid` – maybe that would just get us a tool recommendation for a different HTML prettifier.

I did it time ago (https://github.com/chrisdone/lucid/issues/99) but the maintainer doesn't want to accept that work into that package.

---

I'm trying to use another library that includes a pretty printer for HTML.

**Comment by Gabriel439** [comment]
Yeah, in general it's dangerous to depend on a package that is not properly maintained, for these and other reasons

**Comment by Gabriel439** [comment]
You might also find this useful: http://www.haskellforall.com/2018/05/how-i-evaluate-haskell-packages.html

**Comment by german1608** [comment]
For the record: `blaze-html`, the other option for html generation, does pretty printing

**Comment by german1608** [comment]
I'm noticing that the problem isn't specific to `script` tags: any empty HTML tag found is replaced with their `/>` equivalent.

I found in the golden files the following:

```html
<div />
```

and it should be:

```html
<div></div>
```

The problem is here

https://hackage.haskell.org/package/HaXml-1.25.5/docs/src/Text.XML.HaXml.Html.Pretty.html#element

it doesn't handle properly some HTML closing tags.

From my personal experience, I use the golden tests to have a better diff view when a test fails, and fixing this issue seems complicated at first glance. Also note that even if the pretty-printer was correct, the browser wouldn't render the website correctly because the css, js and svgs are not copied in the data folder, we would need to do that manually.

So I was thinking: If our concern is to see a demo of how the documentation is rendering, Why don't we generate docs on hydra for the golden files? That way we can keep golden files as a utility to see if something is broken in the _content_ of the generated docs

**Comment by german1608** [comment]
> So I was thinking: If our concern is to see a demo of how the documentation is rendering, Why don't we generate docs on hydra for the golden files? That way we can keep golden files as a utility to see if something is broken in the _content_ of the generated docs

On this context, I created #1969 to generate documentation for both Dhall Prelude and test suite package

**Comment by german1608** [comment]
I feel like we don't have to worry about this anymore, or at least in the short term. If somebody wants to develop a feature and they'd like to see how the HTML is generated they could run

    stack run dhall-docs -- --input dhall-docs/tasty/data/package

... to get the generated docs and debug the JS & CSS. On Hydra, we solved this with #1969. Feel free to reopen this issue if I'm missing something

**Comment by sjakobi** [comment]
https://github.com/dhall-lang/dhall-haskell/pull/2156 has replaced `HaXml` with a simplistic prettifier. I believe the HTML files can be viewed in a browser now, although I'm not quite sure whether the prettifier exactly preserves the HTML semantics.
